lib/Runtime/Library/ConcatString.cpp

```diff
@@ -17,11 +17,127 @@ namespace Js
     {
     }
 
+    LiteralStringWithPropertyStringPtr::LiteralStringWithPropertyStringPtr(const char16 * wString,
+      const CharCount stringLength, JavascriptLibrary *const library) :
+        LiteralString(library->GetStringTypeStatic(), wString, stringLength),
+        propertyString(nullptr)
+    {
+    }
+
+    LiteralStringWithPropertyStringPtr *LiteralStringWithPropertyStringPtr::
+    NewFromWideString(const uint16_t * wideString, const CharCount charCount, JavascriptLibrary *const library)
+    {
+        Assert(library != nullptr && wideString != nullptr);
+
+        switch (charCount)
+        {
+            case 0:
+                return (LiteralStringWithPropertyStringPtr*) library->GetEmptyString();
+
+            // TODO: Make CharStringCache LiteralStringWithPropertyStringPtr?
+            // case 1:
+            //     return library->GetCharStringCache().GetStringForChar(*content);
+
+            default:
+                break;
+        }
+
+        Recycler * recycler = library->GetRecycler();
+        ScriptContext * scriptContext = library->GetScriptContext();
+        char16* destString = RecyclerNewArrayLeaf(recycler, WCHAR, charCount + 1);
+
+        if (destString == nullptr)
+        {
+            Js::JavascriptError::ThrowOutOfMemoryError(scriptContext);
+        }
+
+        js_wmemcpy_s(destString, charCount, (LPWSTR)wideString, charCount);
+        destString[charCount] = char16(0);
+
+        return RecyclerNew(library->GetRecycler(), LiteralStringWithPropertyStringPtr, destString, charCount, library);
+    }
+
+    LiteralStringWithPropertyStringPtr* LiteralStringWithPropertyStringPtr::CreateEmptyString(JavascriptLibrary *const library)
+    {
+        return RecyclerNew(library->GetRecycler(), LiteralStringWithPropertyStringPtr, _u(""), 0, library);
+    }
+
+    LiteralStringWithPropertyStringPtr *LiteralStringWithPropertyStringPtr::
+      NewFromCString(const char * cString, const CharCount charCount, JavascriptLibrary *const library)
+    {
+        Assert(library != nullptr && cString != nullptr);
+
+        switch (charCount)
+        {
+            case 0:
+                return (LiteralStringWithPropertyStringPtr*) library->GetEmptyString();
```

Comment:
Could you have GetEmptyString return LiteralStringWithPropertyStringPtr* so we can avoid this cast? otherwise someone could change the emptyString to some other type of string, and compiler won't complain

Reply:
I could put an assert on the returning type (VT check)

Reply:
Sure that would work, but I think it is strictly better to have more type information represented at compile time rather than relying on run time asserts.

Reply:
Aren't we going to cast this type eventually? (library-GetEmptyString)